server/text_generation_server/models/causal_lm.py

```diff
@@ -74,61 +76,113 @@ def round_up(number, k):
     return (number + k - 1) // k * k
 
 
-def prepare_memory(new_bs, tensor, inplace):
-    if inplace:
-        return tensor
-    else:
-        return tensor.new_empty((new_bs,) + tensor.shape[1:])
+def to_tensor_indices(indices, device):
+    return [torch.tensor(idx, dtype=torch.int32, device=device) for idx in indices]
 
 
-def move_data(dst_tensor, chunk_size, indices, src_tensors):
-    batch_dim = 0
-    bs = dst_tensor.size(batch_dim)
-    assert bs % chunk_size == 0, 'Batch dim must be divisible by chunk size!'
-    result = dst_tensor
-    if chunk_size > 1:
-        dst_tensor = dst_tensor.view(bs // chunk_size, chunk_size, *dst_tensor.shape[1:])
-    htorch.core.mark_step()
-    for ind, src_t in zip(indices, src_tensors):
-        if chunk_size > 1:
-            src_t = src_t.view(bs // chunk_size, chunk_size, *src_t.shape[1:])
-        for dst_idx, src_idx in ind:
-            src_data = torch.index_select(src_t, batch_dim, src_idx)
-            dst_tensor.index_copy_(batch_dim, dst_idx, src_data)
-            htorch.core.mark_step()
-    return result
-
-
-def generate_shift_chunks(offset):
-    chunk_sizes = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]
+def calculate_chunks(offset):
     result = []
     while offset != 0:
         sign = 1 if offset > 0 else -1
-        best_chunk = min((abs(offset - sign * c), sign * c) for c in chunk_sizes)[1]
+        best_chunk = min((abs(offset - sign * c), sign * c) for c in CHUNK_SIZES)[1]
         result.append(best_chunk)
         offset = offset - best_chunk
     return result
 
 
-def roll(tensor, dim, chunks):
-    dbg_trace('ROLL', f'shape:{list(tensor.shape)} dim:{dim} chunks:{chunks}')
-    for c in chunks:
-        tensor = torch.roll(tensor, c, dim)
-        htorch.core.mark_step()
-    return tensor
+def biggest_single_chunk(offset):
+    if offset < 0:
+        idx = bisect.bisect(CHUNK_SIZES, -offset)
+        return -CHUNK_SIZES[idx-1]
+    elif offset > 0:
+        idx = bisect.bisect(CHUNK_SIZES, offset)
+        candidate = CHUNK_SIZES[idx]
+        return CHUNK_SIZES[idx-1]
+    else:
+        return 0
 
 
-def shift(tensor, dim, offset):
-    assert dim < 0, 'Only negative dims are supported'
-    if offset == 0:
+def grouped_pad(tensor_groups, dims, values):
+    paddings = [MAX_TOTAL_TOKENS - tensors[0].size(dim) if dim is not None else 0 for tensors, dim in zip(tensor_groups, dims)]
```

Comment:
you iterate over it twice, here and in line 108. I guess you could just simply calculate padding in line 109 as:
`padding = MAX_TOTAL_TOKENS - tensors[0].size(dim) if dim is not None else 0`
both `tensors` and `dim` is available on the loop at line 108.
It will save some host time to trace it

Reply:
Good idea! Thanks!